zquery/src/main/scala/zquery/ZQuery.scala

```diff
@@ -0,0 +1,238 @@
+   */
+  final def provideSome[R0](f: R0 => R): ZQuery[R0, E, A] =
+    new ZQuery[R0, E, A] {
+      def step(cache: Cache): ZIO[R0, E, Result[R0, E, A]] =
+        self.step(cache).provideSome(f).map(_.provideSome(f))
+    }
+
+  /**
+   * Returns an effect that models executing this query.
+   */
+  final val run: ZIO[R, E, A] =
+    runLog.map(_._2)
+
+  /**
+   * Returns an effect that models executing this query with the specified
+   * cache. This can be useful for deterministically "replaying" a query
+   * without executing any new requests.
+   */
+  final def runCache(cache: Cache): ZIO[R, E, A] =
+    step(cache).flatMap {
+      case Result.Blocked(br, c) => br.run *> c.runCache(cache)
+      case Result.Done(a)        => ZIO.succeed(a)
+    }
+
+  /**
+   * Returns an effect that models executing this query, returning the query
+   * result along with the cache containing a complete log of all requests
+   * executed and their results. This can be useful for logging or analysis of
+   * query execution.
+   */
+  final def runLog: ZIO[R, E, (Cache, A)] =
+    for {
+      cache <- Cache.empty
+      a     <- runCache(cache)
+    } yield (cache, a)
+
+  /**
+   * Returns a query that models the execution of this query and the specified
+   * query in parallel, combining their results into a tuple. All requests that
+   * can potentially be executed in parallel will be batched and deduplication
+   * and caching will be applied.
+   */
+  final def zipPar[R1 <: R, E1 >: E, B](that: ZQuery[R1, E1, B]): ZQuery[R1, E1, (A, B)] =
+    zipWithPar(that)((_, _))
+
+  /**
+   * Returns a query that models the execution of this query and the specified
+   * query in parallel, combining their results with the specified function.
+   * All requests that can potentially be executed in parallel will be batched
+   * and deduplication and caching will be applied.
+   */
+  final def zipWithPar[R1 <: R, E1 >: E, B, C](that: ZQuery[R1, E1, B])(f: (A, B) => C): ZQuery[R1, E1, C] =
+    new ZQuery[R1, E1, C] {
+      def step(cache: Cache): ZIO[R1, E1, Result[R1, E1, C]] =
+        self.step(cache).zip(that.step(cache)).map {
+          case (Result.Blocked(br1, c1), Result.Blocked(br2, c2)) => Result.blocked(br1 ++ br2, c1.zipWithPar(c2)(f))
+          case (Result.Blocked(br, c), Result.Done(_))            => Result.blocked(br, c.zipWithPar(that)(f))
+          case (Result.Done(_), Result.Blocked(br, c))            => Result.blocked(br, self.zipWithPar(c)(f))
+          case (Result.Done(a), Result.Done(b))                   => Result.done(f(a, b))
+        }
+    }
+}
+
+object ZQuery {
+
+  /**
+   * Collects a collection of queries into a query returning a collection of
+   * their results. All requests will be batched and deduplication and caching
+   * will be applied.
+   */
+  final def collectAllPar[R, E, A](as: Iterable[ZQuery[R, E, A]]): ZQuery[R, E, List[A]] =
+    foreachPar(as)(identity)
```

Comment:
Can you also add the non-parallel versions for `collectAll` and `foreach`?
In the case of mutations, the spec states that operations have to be done sequentially. I don't expect users to use custom `ZQuery` for mutations, just `IO`, but I internally convert them to `ZQuery` and need to ensure they're not done in parallel.

Reply:
Done.